simulator/func_sim/rf/rf.h

```diff
@@ -36,22 +36,41 @@ class RF
     }
 
     template <typename T>
-    void write( Register num, const T& val, bool accumulating_instr = false)
+    T addition( const T& val)
+    {
+      const auto entry_hi = get_entry( Register::mips_hi);
+      const auto entry_lo = get_entry( Register::mips_lo);
+      auto hi_lo = static_cast<uint64>( entry_hi.value);
+      hi_lo <<= 32;
+      hi_lo += static_cast<uint64>( entry_lo.value);
+      hi_lo += static_cast<uint64>( val);
+      return static_cast<T>( hi_lo);
+    }
+
+    template <typename T>
+    T subtraction( const T& val)
+    {
+      const auto entry_hi = get_entry( Register::mips_hi);
+      const auto entry_lo = get_entry( Register::mips_lo);
+      auto hi_lo = static_cast<uint64>( entry_hi.value);
+      hi_lo <<= 32;
+      hi_lo += static_cast<uint64>( entry_lo.value);
+      hi_lo -= static_cast<uint64>( val);
+      return static_cast<T>( hi_lo);
+    }
+
+    template <typename T>
+    void write( Register num, const T& val, int8 accumulating_instr = 0)
     {
         T new_val = val;
         if ( num.is_zero())
             return;
-        if( accumulating_instr){
-            const auto entry_hi = get_entry( Register::mips_hi);
-            const auto entry_lo = get_entry( Register::mips_lo);
-            auto hi_lo = static_cast<uint64>( entry_hi.value);
-            hi_lo <<= 32;
-            hi_lo += static_cast<uint64>( entry_lo.value);
-            hi_lo += static_cast<uint64>( val);
-            new_val = static_cast<T>( hi_lo);
-        }
+        if ( accumulating_instr == 1)
+            new_val = addition( val);
```

Comment:
If you change signature:
```c++
void write( Register num, T val, int8 accumulating_instr = 0)
```
you may remove `new_val` and use `val = addition( val)` instead